src/provider/bpmn/properties/AdHocCompletionProps.js

```diff
@@ -0,0 +1,159 @@
+import { is, getBusinessObject } from 'bpmn-js/lib/util/ModelUtil';
+
+import {
+  isFeelEntryEdited,
+  CheckboxEntry,
+  isCheckboxEntryEdited,
+} from '@bpmn-io/properties-panel';
+
+import { createElement } from '../../../utils/ElementUtil';
+
+import { useService } from '../../../hooks';
+
+import { FeelEntryWithVariableContext } from '../../../entries/FeelEntryWithContext';
+
+/**
+ * @typedef { import('@bpmn-io/properties-panel').EntryDefinition } Entry
+ */
+
+/**
+ * @returns {Array<Entry>} entries
+ */
+export function AdHocCompletionProps(props) {
+  const { element } = props;
+
+  if (!is(element, 'bpmn:AdHocSubProcess')) {
+    return [];
+  }
+
+  return [
+    {
+      id: 'completionCondition',
+      component: CompletionCondition,
+      isEdited: isFeelEntryEdited,
+    },
+    {
+      id: 'cancelRemainingInstances',
+      component: CancelRemainingInstances,
+      isEdited: isCheckboxEntryEdited,
+    },
+  ];
+}
+
+function CompletionCondition(props) {
+  const { element } = props;
+
+  const bpmnFactory = useService('bpmnFactory');
+  const debounce = useService('debounceInput');
+  const commandStack = useService('commandStack');
+  const translate = useService('translate');
+
+  const getValue = () => {
+    const expression = getBusinessObject(element).get('completionCondition');
+    return expression && expression.get('body');
+  };
+
+  const setValue = (value) => {
+    return commandStack.execute(
+      'element.updateModdleProperties',
+      updateFormalExpression(element, 'completionCondition', value, bpmnFactory)
+    );
+  };
+
+  return FeelEntryWithVariableContext({
+    element,
+    id: 'completionCondition',
+    label: translate('Completion condition'),
+    feel: 'required',
+    getValue,
+    setValue,
+    debounce,
+  });
+}
+
+function CancelRemainingInstances(props) {
+  const { element } = props;
+
+  const commandStack = useService('commandStack');
+  const translate = useService('translate');
+
+  const businessObject = getBusinessObject(element);
+
+  const getValue = () => {
+    return businessObject.get('cancelRemainingInstances');
+  };
+
+  const setValue = (value) => {
+    commandStack.execute('element.updateModdleProperties', {
+      element,
+      moddleElement: businessObject,
+      properties: {
+        cancelRemainingInstances: value,
+      },
+    });
+  };
+
+  return CheckboxEntry({
+    element,
+    id: 'cancelRemainingInstances',
+    label: translate('Cancel remaining instances'),
+    getValue,
+    setValue,
+  });
+}
+
+// helper ////////////////////////////
+
+// formal expression /////////////////
+
+/**
+ * updateFormalExpression - updates a specific formal expression
+ *
+ * @param {djs.model.Base} element
+ * @param {string} propertyName
+ * @param {string} newValue
+ * @param {BpmnFactory} bpmnFactory
+ */
+function updateFormalExpression(element, propertyName, newValue, bpmnFactory) {
```

Comment:
There is a certain amount of duplication across other properties which handle expressions - this one was mostly adapted from the `MultiInstanceProps`. Do we have a preference on duplication vs. extracting into reusable utils/helpers?

Reply:
Given that this is the third place where we can update formal expression (multiinstance, sequence flow condition, ad-hoc subprocess), having a util in https://github.com/bpmn-io/bpmn-js-properties-panel/tree/main/src/utils is a good idea.

Reply:
Refactored into a utility in 68d86a86dd896e8abe80aa03ba9caa38f7a0d3af and adapted the 2 new AdHoc props implementations to use that one. I would create a follow-up PR to adapt the other code parts as well to not expand the scope of this PR.

Reply:
> Refactored into a utility in https://github.com/bpmn-io/bpmn-js-properties-panel/commit/68d86a86dd896e8abe80aa03ba9caa38f7a0d3af and adapted the 2 new AdHoc props implementations to use that one. I would create a follow-up PR to adapt the other code parts as well to not expand the scope of this PR.

As long as you clearly separate refactoring and your improvement in separate commits we're fine. We don't have limits in terms of size of a PR, what we ask for is that the commits are semantically structured. Your changes will also not be squashed, so whatever work you structure will be "followable" later on in the commit history.

Reply:
Fully agree on the semantically structured commits. I'm open to both - the follow-up PR would make it easier to follow the refactoring-only change on unrelated components with a dedicated PR description, but on the other hand changing other components here makes it easier to verify that the abstraction of the helper works properly for multiple use cases 😉 

I now updated the `MultiInstanceProps` to use the new helper in https://github.com/bpmn-io/bpmn-js-properties-panel/pull/1114/commits/533f220dfed13a286d2b9666b94a341b25baea9a but kept the sequence flow as it was as its logic seems a bit different than the other 2.